src/plugins/extra/smartquotes.rs

```diff
@@ -0,0 +1,517 @@
+//! Replaces `"` and `'` quotes with "nicer" ones like `‘`, `’`, `“`, `”`, or
+//! with `’` for words like "isn't".
+//!
+//!
+//! ## Implementation notes
+//!
+//! The main obstacle to implementing this was the fact that the document is
+//! necessarily represented as a tree of nodes.
+//! Each node is thus necessarily referenced by its parents, which means that an
+//! any given moment we cannot hold a mutable reference to a node if any other
+//! part of the code holds a reference to the document. At least that's my
+//! understanding of the problem.
+//! The smartquotes algorithm from the JS library makes heavy use of iteration
+//! backwards and forwards through a flat list of tokens. This isn't really
+//! possible in the Rust implementation. Building a flat representation of all
+//! `Node` objects is easy, but holding that list precludes us from executing a
+//! `root.walk_mut` call at the same time.
+//! On top of that, while the smartquotes algorithm iterates linearly over all
+//! nodes/tokens, looking at a specific token with index `j` can trigger
+//! replacements in any of the tokens with `0` to `j - 1`.
+//!
+//! The solution proposed here is to first compute all the replacement
+//! operations on a read-only flat view of the document, and _then_ to perform
+//! all replacements in a single call to `root.walk_mut`.
+use crate::parser::core::CoreRule;
+use crate::parser::inline::Text;
+use crate::plugins::cmark::block::paragraph::Paragraph;
+use crate::plugins::cmark::inline::newline::{Hardbreak, Softbreak};
+use crate::plugins::html::html_inline::HtmlInline;
+use crate::{MarkdownIt, Node};
+use once_cell::sync::Lazy;
+use regex::Regex;
+use std::collections::HashMap;
+
+const APOSTROPHE: char = '\u{2019}';
+const SINGLE_QUOTE: char = '\'';
+const DOUBLE_QUOTE: char = '"';
+const SPACE: char = ' ';
+
+static PUNCTUATION_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"\p{Punctuation}").unwrap());
+
+pub fn add(md: &mut MarkdownIt) {
+    md.add_rule::<SmartQuotesRule<'‘', '’', '“', '”'>>();
+}
```

Comment:
probably want to add an ability for user to configure it, i.e. smth like this:

```rs
pub fn add(md: &mut MarkdownIt) {
    add_with<'‘', '’', '“', '”'>();
}

pub fn add_with<const A: char, const B: char, const C: char, const D: char>(md: &mut MarkdownIt) {
    md.add_rule::<SmartQuotesRule<A, B, C, D>>();
}
```

configurable quotes are a feature in markdown-it, nice to see it here... the sheer quantity of generic arguments look very weird though